### PR TITLE
Design proposal: secure registry support

### PR DESCRIPTION
**Pull request by GeekArthur**
Signed-off-by: jingfu wang <jingfu.j.wang@ibm.com>

**What type of PR is this?**
/kind documentation
[skip ci]

**Which issue(s) this PR fixes**:
Parent issue: https://github.com/openshift/odo/issues/2893

### CODE REVIEW

**Comment by GeekArthur** [comment]
> Looks good. It looks like you have addressed my early comments on the design from our previous conversation already. One thing that I think is missing is slightly beyond the catalog list, when someone is doing `odo create` on a component that belongs to a secure registry, we'll need to make sure we handle the component creation properly (including download of the devfile from the secure repo) using same secure setting.

Thanks for reviewing! I add the `Create devfile component from secure registry` section to the design proposal.

File: docs/proposals/secure-registry.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+Context:
+1. Given GitHub is going to depreciate basic authentication with username/password https://developer.github.com/changes/2020-02-14-deprecating-password-auth/, we have to only support personal access token authentication method for GitHub-hosted registry scenario.
+2. For cluster-hosted registry, the registry architecture would be creating a NGINX server to host the secure registry, then create a ingress gateway for the NGNIX server to let client access.
+
+Work flow to access secure registry:
+1. Collect Credential
+
+    Currently we have `odo registry add <registry name> <registry URL>` and `odo registry update <registry name> <registry URL>` to add and update registry accordingly. Regarding the CLI design, we can implement the following CLI design to support collecting credential:
+    - `odo registry add <registry name> <registry URL> --token <token>`
```
Comment (kadel) [inline docs/proposals/secure-registry.md:42]: What would be a command for adding registry with basic-auth?

I can see two options:

`odo registry add <registry name> <registry URL> --user <user> --password <password>`
or 
`odo registry add <registry name> <registry URL> --basic-auth <user>:<password>`

Comment (GeekArthur) [reply docs/proposals/secure-registry.md:42]: I like the first option as it aligns with `kubectl`, `docker`, `oc` and `odo`, I am gonna update the design proposal.

**Comment by GeekArthur** [comment]
@elsony @kadel Comments addressed, please review again, thanks

**Comment by gorkem** [comment]
Also consider https://github.com/devfile/kubernetes-api/issues/82

**Comment by GeekArthur** [comment]
> Also consider [devfile/kubernetes-api#82](https://github.com/devfile/kubernetes-api/issues/82). I would hold a decision until it is concluded on the devfile side. If we decide on the OCI based registry it is possible that some (all?) of the concerns on this document will be addressed.

@gorkem Thanks for sharing the OCI-based registry architecture design, this design proposal only focuses on secure registry support, more specifically it only addresses concerns of authentication and authorization stages for existing supported registry and it's not related with the registry design. We can open separate issue and create another design proposal for the registry design to adopt the architecture of OCI-based registry.

File: docs/proposals/secure-registry.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    req.Header.Add("Authorization", bearer)
+    ```
+
+4. Delete Credential
+
+    If multiple secure registries share the same credential, `odo registry delete <registry name>` will delete the credential from keyring instance once the last secure registry using that credential has been deleted.
+
+Create devfile component from secure registry:
+
+When downloading devfile from secure registry, we validate if the credential is valid by adding token to the request header and checking the response.
```
Comment (adisky) [inline docs/proposals/secure-registry.md:73]: we are not doing any validation on `odo registry add <registry name> <registry URL> --user <user> --password <password>`? Just thinking out here, shouldn't we validate credentials and url here? than doing at the time of `odo create`?

Comment (GeekArthur) [reply docs/proposals/secure-registry.md:73]: Regarding arguments/flags validation, the design is that we divide the validation into two stages:
Early validation stage: Only validate URL basic format, happens when user adds registry URL
Validation stage: Validate URL accessibility, happens when user uses registry URL

That design is based on `yum-config-manager`, design proposal with details under this thread: https://github.com/openshift/odo/pull/2940#discussion_r417440866

**Pull request merged**